iac-extensions/docker/src/test/java/org/sonar/iac/docker/visitors/DockerSymbolVisitorTest.java

```diff
@@ -0,0 +1,240 @@
+
+class DockerSymbolVisitorTest {
+
+  private static final String IMAGE_PREFIX = "FROM foobar\n";
+  private final InputFileContext inputFileContext = mock(InputFileContext.class);
+
+  @Test
+  void argInstructionShouldCreateSymbol() {
+    DockerImage image = scanImage("ARG foo=bar");
+    NewKeyValuePair keyValuePair = firstDescendant(image, NewKeyValuePair.class);
+
+    Scope scope = image.scope();
+
+    assertThat(scope.getSymbols()).hasSize(1);
+    assertThat(image.scope().getSymbol("bar")).isNull();
+
+    Symbol symbol = image.scope().getSymbol("foo");
+    assertThat(symbol).isNotNull();
+    assertThat(symbol.usages()).allSatisfy(usage -> {
+      assertThat(usage.kind()).isEqualTo(Usage.Kind.ASSIGNMENT);
+      assertThat(usage.tree()).isEqualTo(keyValuePair);
+    });
+  }
+
+  @Test
+  void argInstructionShouldCreateTwoSymbols() {
+    DockerImage image = scanImage("ARG foo1=bar foo2=bar");
+    List<NewKeyValuePair> keyValuePairs = firstDescendant(image, ArgInstruction.class).keyValuePairs();
+
+    Symbol symbol1 = image.scope().getSymbol("foo1");
+    assertThat(symbol1.usages()).extracting(Usage::tree).containsExactly(keyValuePairs.get(0));
+
+    Symbol symbol2 = image.scope().getSymbol("foo2");
+    assertThat(symbol2.usages()).extracting(Usage::tree).containsExactly(keyValuePairs.get(1));
+  }
+
+  @Test
+  void scopeShouldBeInherit() {
+    Body body = scanBody("ARG foo=bar\n" + IMAGE_PREFIX);
+    Scope bodyScope = body.scope();
+    Scope imageScope = body.dockerImages().get(0).scope();
+
+    assertThat(bodyScope).isNotEqualTo(imageScope);
+    assertThat(bodyScope.getSymbol("foo")).isNotNull();
+    assertThat(imageScope.getSymbol("foo")).isNotNull();
+    assertThat(bodyScope.getSymbol("foo"))
+      .isNotSameAs(imageScope.getSymbol("foo"))
+      .usingRecursiveComparison()
+      .isEqualTo(imageScope.getSymbol("foo"));
+  }
+
+  @Test
+  void twoArgInstructionShouldCreateTwoSymbols() {
+    DockerImage image = scanImage( "ARG foo1=bar\nARG foo2=bar");
+    assertThat(image.scope().getSymbols())
+      .extracting(Symbol::name).containsExactly("foo1", "foo2");
+  }
+
+  @Test
+  void bodyScopeShouldNotBeAffectedByImage() {
+    Body body = scanBody( IMAGE_PREFIX +"ARG foo=bar");
+    Scope bodyScope = body.scope();
+    Scope imageScope = body.dockerImages().get(0).scope();
+
+    assertThat(bodyScope.getSymbol("foo")).isNull();
+    assertThat(imageScope.getSymbol("foo")).isNotNull();
+  }
+
+  @Test
+  void bodyScopeShouldNotBeOverrideByImage() {
+    Body body = scanBody( "ARG foo=bar\n" + IMAGE_PREFIX + "ARG foo=bar");
+    Scope bodyScope = body.scope();
+    Scope imageScope = body.dockerImages().get(0).scope();
+
+    assertThat(bodyScope.getSymbol("foo")).isNotNull()
+      .satisfies(symbol -> assertThat(symbol.usages()).hasSize(1));
+
+    assertThat(imageScope.getSymbol("foo")).isNotNull()
+      .satisfies(symbol -> assertThat(symbol.usages()).hasSize(2));
+  }
```

Comment:
I would add a check to the value, and I would add another test case :
```
ARG foo=bar
FROM scratch
ARG foo
```
For now I guess the second `foo` will have no value, but later when we will support it we will expect it to have `bar` for value.

And for later, other test cases :
```
ARG foo=bar
FROM scratch
ARG foo
LABEL my_label=$foo
# assert the value is bar
```
and
```
ARG foo=bar
FROM scratch
LABEL my_label=$foo
# assert the value is empty string + no symbol created for the $foo in the image
```